File: Python/10.regularExpressionMatching.py
class Solution(object):
    def isMatch(self, s, p):
        memo = {}
        def dp(i, j):
            if (i, j) not in memo:
                if j == len(p):
                    ans = i == len(s)
                else:
                    match = i < len(s) and p[j] in {s[i], '.'}
                    if j+1 < len(p) and p[j+1] == '*':
                        ans = dp(i, j+2) or match and dp(i+1, j)
                    else:
                        ans = match and dp(i+1, j+1)
                memo[i, j] = ans
            return memo[i, j]
        return dp(0, 0)

File: Javascript/10.regularExpressionMatching.js
/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
var isMatch = function(s, p) {
    if(p == '' || !p) {
        return s == '';
    }
    let match = (s && (s.charAt(0) == p.charAt(0) || p.charAt(0) == '.'));
    if (p.length >= 2 && p.charAt(1) == '*') {
        return isMatch(s, p.substring(2)) || (match && isMatch(s.substring(1), p))
    }
    else {
        return match && isMatch(s.substring(1), p.substring(1))
    }
};